### PR TITLE
Add CMake build system

### PR DESCRIPTION
**Pull request by xxks-kkk**
This pull request contains the following changes:

- Reorganize the code repo to support both CMake build system and Autotools system
- Add CMake build system
- Fix `corruption_test` and `table_test` (in order to let CMake compile successfully)
- Add `demo/installation_test` to verify the success of the installation

Verification:

```
mkdir -p build && cd build
cmake .. && make install -j16
make db_test
./db_test
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,17 +47,29 @@ PebblesDB requires `libsnappy` and `libtool`. To install on Linux, please use
...
-`$ cd pebblesdb/`  
-`$ autoreconf -i`  
-`$ ./configure`  
-`$ make`  
-`# make install`  
-`# ldconfig`  
+
+Using autotools:
+
+```shell
```
Comment (vijay03) [inline README.md]: not sure the shell line is required. Could you remove?

File: README.md [hunk 2]
```diff
@@ -125,6 +137,14 @@ ___
   2). Setting this parameter high will favor write throughput while
   setting it lower will favor read/seek throughputs.
 
+---
+### Improvements made after the SOSP paper
+
+The following improvements are made to the codebase after the SOSP paper:
+
+- Add CMake build system support
```
Comment (vijay03) [inline README.md]: Please add your github id in here :)

File: src/AUTHORS [hunk 3]
```diff
@@ -17,3 +17,4 @@ Robert Escriva <robert@hyperdex.org>
 # PebblesDB authors:
 Pandian Raju <pandian@cs.utexas.edu>
 Vijay Chidambaram <vijay@cs.utexas.edu>
+Zeyuan Hu <iamzeyuanhu@utexas.edu>
```
Comment (vijay03) [inline src/AUTHORS]: Could you also add Jayashree Mohan (jaya@cs.utexas.edu)

File: .gitignore [hunk 4]
```diff
@@ -67,3 +67,16 @@
...
+.vscode/
+cmake-build-debug/
+testPebblesDB
+gdb.txt
+.autotools
+.cproject
+.project
+.settings/
+nbproject/
+CMakeFiles
```
Comment (vijay03) [inline .gitignore]: Add a newline at the end please

File: README.md (hunk 1 above)
Comment (xxks-kkk) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (xxks-kkk) [reply README.md]: done 😄

File: src/AUTHORS (hunk 3 above)
Comment (xxks-kkk) [reply src/AUTHORS]: done

File: .gitignore (hunk 4 above)
Comment (xxks-kkk) [reply .gitignore]: done

**Comment by vijay03** [comment]
Hi,

Thanks again for your work on this!  I tried locally testing the pull request on my Mac OSX 10.11.6 environment, and it wouldn't compile. I'm not sure if I'm missing new dependencies.
[CMakeError.log](https://github.com/utsaslab/pebblesdb/files/2251537/CMakeError.log)
[CMakeOutput.log](https://github.com/utsaslab/pebblesdb/files/2251538/CMakeOutput.log)

I'll look into this again in a bit, but it would be helpful if you could look and see if I'm missing something since PebblesDB used to compile previously for me.

**Comment by xxks-kkk** [comment]
Hi Vijay,

From the error log you posted, I see there are several header files cannot be located on your system

```
/Users/vijay/work/pebblesdb/build/CMakeFiles/CMakeTmp/CheckIncludeFile.c:1:10: fatal error: 'endian.h' file not found
#include <endian.h>
         ^
1 error generated.
```
Those headers include `endian.h`, `sys/endian.h`, `sys/isa_defs.h`. I'm investigating why this is the case. Meanwhile, can you please paste the output of `cmake ..`? Also, Can you also help to confirm that there is no such problem when building with Autotools?

Thanks!

**Comment by vijay03** [comment]
I've attached the cmake output in the previous comment. I can confirm I can successfully build the master branch.

**Comment by xxks-kkk** [comment]
Sorry for my bad wording. Can you please paste the terminal output of the command `cmake ..` . Also, I check with my local "CMakeError.log", the missing header is not the root cause of the problem (i.e., I have several headers missing as well but can still build).  I guess the terminal output of the commands `cmake ..` and `make all -j4` will help in this case.

**Comment by xxks-kkk** [comment]
Hello Vijay,

I forget to push one more file that is required by the `CMake` (got ignored due to my `.gitignore`). I push another commit. Please try again and let me know how it goes.

Thanks!

Zack

**Comment by vijay03** [comment]
Great! It works for me now. I was able to pass PebblesDB tests successfully on my machine.

**Pull request merged**